### DISCUSSION TITLE
Error: WebSocket implementation missing

### DISCUSSION THREAD

**Discussion by xndyz** [enisdenjo/graphql-ws#315] [Q&A] [answered]
**Screenshot**
<img width="1040" alt="CleanShot 2022-02-21 at 06 02 28@2x" src="https://user-images.githubusercontent.com/19108762/154942779-376ce1c3-f184-4b77-ae39-8931f1af45eb.png">

Visualising is always helpful.

**Expected Behaviour**
I expected it to do _this_ -
No error

**Actual Behaviour**
but instead it did _this_.
Shows this error until you refresh the page.

**Debug Information**
Help us debug the bug?
This app communicates with Hasura Cloud via WS. I am setting up the client like so:

```
const client = createClient({
  url: process.env.NEXT_PUBLIC_GRAPHQL_WS,

  connectionParams: () => {
    const token = Cookies.get("token");
    if (!token) {
      return {};
    }

    return {
      Authorization: `Bearer ${token}`,
    };
  },
});
```

**Comment by enisdenjo** [comment, marked as answer]
The error is self-explanatory, WebSocket implementation is missing, you aren't running the client in the browser. Please consult the ["Client usage in Node" recipe](https://github.com/enisdenjo/graphql-ws#node-client) and provide the [`webSocketImpl`](https://github.com/enisdenjo/graphql-ws/blob/master/docs/interfaces/client.ClientOptions.md#websocketimpl) argument.

Or alternatively, if you're using SSR, see https://github.com/enisdenjo/graphql-ws/discussions/215#discussioncomment-1032087.

**Reply by xndyz** [reply]
I get this error after using the `ws` module as `webSocketImpl`:

<img width="988" alt="CleanShot 2022-02-21 at 14 49 46@2x" src="https://user-images.githubusercontent.com/19108762/155018138-4566a24d-bf75-4bc6-9e5b-50e9e31782a9.png">

**Reply by xndyz** [reply]
You think this has to do with SSR in Next.js?

**Reply by enisdenjo** [reply]
Yes, please see also https://github.com/enisdenjo/graphql-ws/discussions/215#discussioncomment-1032087.

**Reply by xndyz** [reply]
Thanks! That fixed it.

**Reply by 61ue5peed** [reply]
Yep: !!! 
```
import { w3cwebsocket } from 'websocket';
...
const myWSClient =  createClient({
  url: websockerUrl || "",
  webSocketImpl: w3cwebsocket,
});
const wsLink = new GraphQLWsLink(myWSClient);
```
